src/dawn/CodeGen/Cuda/CudaCodeGen.cpp

```diff
@@ -0,0 +1,1139 @@
+  iir::MultiInterval mInterval{computePartitionOfIntervals(ms)};
+  return mInterval.contiguous() && (ms->getLoopOrder() == iir::LoopOrderKind::LK_Parallel);
+}
+
+iir::Interval::IntervalLevel
+CudaCodeGen::computeNextLevelToProcess(const iir::Interval& interval,
+                                       iir::LoopOrderKind loopOrder) const {
+  iir::Interval::IntervalLevel intervalLevel;
+  if(loopOrder == iir::LoopOrderKind::LK_Backward) {
+    intervalLevel = interval.upperIntervalLevel();
+  } else {
+    intervalLevel = interval.lowerIntervalLevel();
+  }
+  return intervalLevel;
+}
+void CudaCodeGen::generateIJCacheIndexInit(MemberFunction& kernel,
+                                           const CacheProperties& cacheProperties,
+                                           const Array3ui blockSize) const {
+  if(cacheProperties.isThereACommonCache()) {
+    kernel.addStatement("int " +
+                        cacheProperties.getCommonCacheIndexName(iir::Cache::CacheTypeKind::IJ) +
+                        "= iblock + " + std::to_string(cacheProperties.getOffsetCommonCache(1)) +
+                        " + (jblock + " + std::to_string(cacheProperties.getOffsetCommonCache(1)) +
+                        ")*" + std::to_string(cacheProperties.getStrideCommonCache(1, blockSize)));
+  }
+}
+
+bool CudaCodeGen::useTmpIndex(
+    const std::unique_ptr<iir::MultiStage>& ms,
+    const std::shared_ptr<iir::StencilInstantiation>& stencilInstantiation) const {
+  const auto& fields = ms->getFields();
+  const bool containsTemporary =
+      (find_if(fields.begin(), fields.end(), [&](const std::pair<int, iir::Field>& field) {
+         const int accessID = field.second.getAccessID();
+         // we dont need to initialize tmp indices for fields that are cached
+         bool cacheBypass = accessIsCached(accessID, ms);
+         return stencilInstantiation->isTemporaryField(accessID) && !cacheBypass;
+       }) != fields.end());
+
+  return containsTemporary;
+}
+
+bool CudaCodeGen::accessIsCached(const int accessID,
+                                 const std::unique_ptr<iir::MultiStage>& ms) const {
+  return ms->isCached(accessID) &&
+         (ms->getCache(accessID).getCacheType() == iir::Cache::CacheTypeKind::IJ);
+}
+
+void CudaCodeGen::generateTmpIndexInit(
+    MemberFunction& kernel, const std::unique_ptr<iir::MultiStage>& ms,
+    const std::shared_ptr<iir::StencilInstantiation>& stencilInstantiation) const {
+
+  if(!useTmpIndex(ms, stencilInstantiation))
+    return;
+
+  auto maxExtentTmps = computeTempMaxWriteExtent(*(ms->getParent()));
+  kernel.addStatement("int idx_tmp = (iblock+" + std::to_string(-maxExtentTmps[0].Minus) +
+                      ")*1 + (jblock+" + std::to_string(-maxExtentTmps[1].Minus) + ")*jstride_tmp");
+}
+
+int CudaCodeGen::paddedBoundary(int value) {
+  return value <= 1 ? 1 : value <= 2 ? 2 : value <= 4 ? 4 : 8;
+}
+void CudaCodeGen::generateAllCudaKernels(
```

Comment:
why put this into a fctn?

Reply:
well, this is three lines and only called once... there is more code due to this being a function